File: libcompositor/source/mtl2/CpsrShaderFunction.m
#include "CpsrGraphics+Private.h"

// clang-format off
static inline MTLLanguageVersion ShaderVersionAsMetalType(CpsrShaderType shaderType) {
  switch (shaderType & CPSR_SHADER_VERSION_MASK) {
#if TARGET_OS_IPHONE
  case CPSR_SHADER_VERSION_MSL1_0:  return MTLLanguageVersion1_0;
#endif
  case CPSR_SHADER_VERSION_MSL2_0:  return MTLLanguageVersion2_0;
  case CPSR_SHADER_VERSION_MSL2_1:  return MTLLanguageVersion2_1;
  case CPSR_SHADER_VERSION_MSL2_2:  return MTLLanguageVersion2_2;
  case CPSR_SHADER_VERSION_MSL1_1:
  default:                          return MTLLanguageVersion1_1;
  }
}
// clang-format on

static inline CpsrShaderFunction *ShaderFunctionCreate(id<MTLFunction> native) {
  CpsrShaderFunction *function = CpsrAlloc(CpsrShaderFunction);
  if (function) {
    function->native = native;
  }
  return function;
}

CpsrShaderFunction *CpsrShaderFunctionCreateFromLibrary(const CpsrShaderLibrary *shaderLibrary, const char *functionName) {
  NSString *nsFunctionName = [[NSString alloc] initWithBytesNoCopy:(void *)functionName length:strlen(functionName) encoding:NSUTF8StringEncoding freeWhenDone:NO];
  id<MTLFunction> native = [shaderLibrary->native newFunctionWithName:nsFunctionName];
  
  [nsFunctionName release];
  
  if (native) {
    CpsrShaderFunction *function = ShaderFunctionCreate(native);
    return function;
  }
  return NULL;
}

// TODO: Macro support
CpsrShaderFunction *CpsrShaderFunctionCreateFromString(const CpsrDevice *device, const char *functionString, CpsrShaderType shaderType) {
  if ((shaderType & CPSR_SHADER_LANGUAGE_MASK) != CPSR_SHADER_MSL) {
    return NULL;
  }
  
  NSString *nsFunctionString = [[NSString alloc] initWithBytesNoCopy:(void *)functionString length:strlen(functionString) encoding:NSUTF8StringEncoding freeWhenDone:NO];
  
  MTLCompileOptions *options = [MTLCompileOptions new];
  options.languageVersion = ShaderVersionAsMetalType(shaderType);
  
  NSError *error = nil;
  id<MTLLibrary> library = [device->native newLibraryWithSource:nsFunctionString options:options error:&error];
  
  [options release];
  [nsFunctionString release];
  
  if (library) {
    id<MTLFunction> native = [library newFunctionWithName:@"_main"];
    if (native) {
      CpsrShaderFunction *function = ShaderFunctionCreate(native);
      return function;
    }
  }
  
  if (error) {
    // TODO: Log
  }
  return NULL;
}

void CpsrShaderFunctionDestroy(CpsrShaderFunction *shaderFunction) {
  CPSR_ASSUME(shaderFunction);
  
  [shaderFunction->native release];
  CpsrDealloc(shaderFunction);
}

File: libcompositor/source/dx12/CpsrShaderFunction.c
#include "CpsrGraphics+Private.h"

// clang-format off
static inline void ShaderTypeAsDx12Type(CpsrShaderType shaderType, char out[7]) {
  switch (shaderType & CPSR_SHADER_TYPE_MASK) {
  case CPSR_SHADER_TYPE_PIXEL_SHADER:   out[0] = 'p'; break;
  case CPSR_SHADER_TYPE_COMPUTE_SHADER: out[0] = 'c'; break;
  case CPSR_SHADER_TYPE_VERTEX_SHADER:
  default:                              out[0] = 'v'; break;
  }
}

static inline void ShaderVersionAsDx12Type(CpsrShaderType shaderType, char out[7]) {
  switch (shaderType & CPSR_SHADER_VERSION_MASK) {
  case CPSR_SHADER_VERSION_HLSL5_1:  memcpy(out, "xs_5_1", 7); break;
  case CPSR_SHADER_VERSION_HLSL6_0:  memcpy(out, "xs_6_0", 7); break;
  case CPSR_SHADER_VERSION_HLSL6_1:  memcpy(out, "xs_6_1", 7); break;
  case CPSR_SHADER_VERSION_HLSL5_0:
  default:                           memcpy(out, "xs_5_0", 7); break;
  }
}
// clang-format on

#include <d3dcompiler.h>

CpsrShaderFunction *CpsrShaderFunctionCreateFromLibrary(const CpsrShaderLibrary *shaderLibrary,
                                                        const char *functionName) {
  size_t filepathLength = shaderLibrary->filepathLength;
  size_t filenameLength = filepathLength + strlen(functionName);
  char *savedFilename = (char *)malloc(sizeof(char) * (filenameLength + 1));
  memcpy(savedFilename, shaderLibrary->filepath, filepathLength + 1);
  strcat(savedFilename, functionName);

  ID3DBlob *native;
  HRESULT hr = D3DReadFileToBlob(savedFilename, &native);
  free(savedFilename);
  if (FAILED(hr)) {
    return NULL;
  }

  CpsrShaderFunction *function = CpsrAlloc(CpsrShaderFunction);
  if (function) {
    function->native = native;
  }
  return function;
}

// TODO: Macro support
CpsrShaderFunction *CpsrShaderFunctionCreateFromString(const CpsrDevice *device,
                                                       const char *functionString,
                                                       CpsrShaderType shaderType) {
  if ((shaderType & CPSR_SHADER_LANGUAGE_MASK) != CPSR_SHADER_HLSL) {
    return NULL;
  }

  char target[7];
  ShaderVersionAsDx12Type(shaderType, target);
  ShaderTypeAsDx12Type(shaderType, target);

  UINT flag = D3DCOMPILE_ENABLE_STRICTNESS;
#ifndef NDEBUG
  flag = D3DCOMPILE_DEBUG | D3DCOMPILE_OPTIMIZATION_LEVEL0;
#else
  flag = D3DCOMPILE_OPTIMIZATION_LEVEL3;
#endif

  ID3DBlob *native;
  ID3DBlob *errorBlob;
  HRESULT hr = D3DCompile(
      functionString, strlen(functionString), NULL, NULL, NULL, "_main", target, flag, 0, &native, &errorBlob);
  if (FAILED(hr)) {
    // TODO: error log
    return NULL;
  }

  CpsrShaderFunction *function = CpsrAlloc(CpsrShaderFunction);
  if (function) {
    function->native = native;
  }
  return function;
}

void CpsrShaderFunctionDestroy(CpsrShaderFunction *shaderFunction) {
  CPSR_ASSUME(shaderFunction);

  ID3D10Blob_Release(shaderFunction->native);
  CpsrDealloc(shaderFunction);
}
